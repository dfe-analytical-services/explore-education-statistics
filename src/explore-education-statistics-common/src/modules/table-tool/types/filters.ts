/* eslint-disable max-classes-per-file */
import {
  FilterOption,
  GeoJsonFeature,
  IndicatorOption,
  LocationOption,
  TimePeriodOption,
} from '@common/services/tableBuilderService';

interface GroupedFilterOption extends FilterOption {
  group?: string;
}

export abstract class Filter {
  public readonly value: string;

  public readonly label: string;

  public readonly group?: string;

  protected constructor({ value, label, group }: GroupedFilterOption) {
    this.value = value;
    this.label = label;
    this.group = group;
  }

  /**
   * Get an identifier that guarantees
   * uniqueness for the filter somehow.
   *
   * This is unlike {@property value} where
   * duplicates are allowed (e.g. locations),
   * making it potentially unsafe for equality
   * checks with other filters of the same type.
   */
  public get id(): string {
    return this.value;
  }
}

export class CategoryFilter extends Filter {
  public readonly isTotal: boolean;

  public readonly category: string;

  public constructor({
    value,
    label,
    group,
    isTotal = false,
    category,
  }: GroupedFilterOption & { isTotal?: boolean; category: string }) {
    super({ value, label, group });
    this.isTotal = isTotal;
    this.category = category;
  }
}

export interface LocationCompositeId {
  level: string;
  value: string;
}

export class LocationFilter extends Filter {
  public readonly level: string;

  public readonly geoJson?: GeoJsonFeature[];

  public constructor({
    value,
    label,
    level,
    geoJson,
    group,
  }: GroupedFilterOption & LocationOption) {
    super({ value, label, group });

    this.level = level;
    this.geoJson = geoJson;
  }

  public static createId(params: { value: string; level: string }): string {
    return JSON.stringify({
      level: params.level,
      value: params.value,
    });
  }

  public static parseCompositeId(id: string): LocationCompositeId {
    const { level, value } = JSON.parse(id);

    if (typeof level !== 'string' || typeof value !== 'string') {
      throw new Error(`Could not parse invalid location composite id: ${id}`);
    }

    return {
      level,
      value,
    };
  }

  public get id(): string {
    return LocationFilter.createId({
      value: this.value,
      level: this.level,
    });
  }
}

export class Indicator extends Filter {
  public readonly unit: string;

  public readonly decimalPlaces?: number;

  public readonly name: string;

  public constructor({
    value,
    label,
    unit,
    group,
    name,
    decimalPlaces,
  }: GroupedFilterOption & IndicatorOption) {
    super({ value, label, group });

    this.unit = unit;
    this.name = name;
    this.decimalPlaces = decimalPlaces;
  }
}

export class TimePeriodFilter extends Filter {
  public readonly year: number;

  public readonly code: string;

  public readonly order: number;

  public constructor({
    year,
    code,
    label,
    order,
  }: TimePeriodOption & { order: number }) {
    super({ label, value: `${year}_${code}` });

    this.code = code;
    this.year = year;
    this.order = order;
  }
}
