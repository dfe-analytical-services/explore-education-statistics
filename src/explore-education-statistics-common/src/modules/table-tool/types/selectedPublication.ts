// TODO: EES-4312 Cleanup this type - use PublicationTreeSummary
export interface SelectedPublication {
  id: string;
  title: string;
  slug: string;
  selectedRelease: SelectedRelease;
  latestRelease: {
    title: string;
  };
}

// TODO: EES-4312 Cleanup this type - use PublicationReleaseSummary
export interface SelectedRelease {
  id: string;
  slug: string;
  latestData: boolean;
  title: string;
}
