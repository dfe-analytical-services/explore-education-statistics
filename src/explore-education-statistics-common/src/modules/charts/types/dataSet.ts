import { ChartSymbol, LineStyle } from '@common/modules/charts/types/chart';
import {
  CategoryFilter,
  Filter,
  Indicator,
  LocationCompositeId,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import { Dictionary } from '@common/types';

export interface ChartData {
  name: string;
  [key: string]: number | string;
}

export interface DataSet {
  indicator?: string;
  filters: string[];
  location?: LocationCompositeId;
  order?: number;
  timePeriod?: string;
}

/**
 * Expanded variant of a data set where
 * we have all the filter details available.
 */
export interface ExpandedDataSet {
  indicator: Indicator;
  filters: CategoryFilter[];
  location?: LocationFilter;
  timePeriod?: TimePeriodFilter;
}

export interface DataSetCategory {
  filter: Filter;
  dataSets: Dictionary<{
    dataSet: DataSetConfiguration;
    value: number;
  }>;
}

/**
 * Don't use this for storing configuration.
 * @deprecated use {@see LegendItem} instead.
 */
export interface DataSetConfigurationOptions {
  label: string;
  colour: string;
  symbol?: ChartSymbol;
  lineStyle?: LineStyle;
}

/**
 * Don't use this for storing configuration.
 * @deprecated use {@see LegendItem} instead.
 */
export interface DataSetConfiguration extends DataSet {
  config?: DataSetConfigurationOptions;
}
