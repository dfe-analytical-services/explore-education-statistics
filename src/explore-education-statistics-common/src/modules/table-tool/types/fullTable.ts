import {
  CategoryFilter,
  Indicator,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import {
  BoundaryLevel,
  TableDataResult,
} from '@common/services/tableBuilderService';
import { Dictionary } from '@common/types';

export interface FullTableMeta {
  publicationName: string;
  subjectName: string;
  locations: LocationFilter[];
  timePeriodRange: TimePeriodFilter[];
  filters: Dictionary<{
    name: string;
    options: CategoryFilter[];
  }>;
  indicators: Indicator[];
  boundaryLevels: BoundaryLevel[];
  footnotes: {
    id: string;
    label: string;
  }[];
  geoJsonAvailable: boolean;
}

export interface FullTable {
  title?: string;
  subjectMeta: FullTableMeta;
  results: TableDataResult[];
}
