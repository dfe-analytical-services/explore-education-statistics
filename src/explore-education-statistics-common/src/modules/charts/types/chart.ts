import { horizontalBarBlockDefinition } from '@common/modules/charts/components/HorizontalBarBlock';
import { infographicBlockDefinition } from '@common/modules/charts/components/InfographicBlock';
import { lineChartBlockDefinition } from '@common/modules/charts/components/LineChartBlock';
import { mapBlockDefinition } from '@common/modules/charts/components/MapBlock';
import { verticalBarBlockDefinition } from '@common/modules/charts/components/VerticalBarBlock';
import { DataSet } from '@common/modules/charts/types/dataSet';
import { LegendConfiguration } from '@common/modules/charts/types/legend';
import { FullTableMeta } from '@common/modules/table-tool/types/fullTable';
import { TableDataResult } from '@common/services/tableBuilderService';
import { NestedPartial } from '@common/types';

export type ChartType =
  | 'line'
  | 'verticalbar'
  | 'horizontalbar'
  | 'map'
  | 'infographic';

export type ChartSymbol =
  | 'circle'
  | 'cross'
  | 'diamond'
  | 'square'
  | 'star'
  | 'triangle'
  | 'wye'
  | 'none';

export type LineStyle = 'solid' | 'dashed' | 'dotted';

export type Axis = 'x' | 'y';
export type AxisGroupBy = 'timePeriod' | 'locations' | 'filters' | 'indicators';
export type AxisType = 'major' | 'minor';
export type TickConfig = 'default' | 'startEnd' | 'custom';
export type ReferenceLineStyle = 'dashed' | 'solid' | 'none';
export type LineChartDataLabelPosition = 'above' | 'below';
export type BarChartDataLabelPosition = 'inside' | 'outside';

export interface ReferenceLine {
  label: string;
  otherAxisPosition?: number;
  position: number | string;
  style?: ReferenceLineStyle;
}

export interface Label {
  text: string;
  rotated?: boolean;
  width?: number;
}

export interface AxisConfiguration {
  type: AxisType;
  groupBy?: AxisGroupBy;
  groupByFilter?: string;
  sortBy?: string;
  sortAsc?: boolean;
  dataSets: DataSet[];
  referenceLines: ReferenceLine[];
  visible: boolean;
  unit?: string;
  showGrid?: boolean;
  label?: Label;
  size?: number;
  min?: number;
  max?: number;
  tickConfig?: TickConfig;
  tickSpacing?: number;
}

export type AxesConfiguration = {
  [key in AxisType]?: AxisConfiguration;
};

export const dataGroupingTypes = {
  EqualIntervals: 'Equal intervals',
  Quantiles: 'Quantiles',
  Custom: 'Custom',
} as const;

export type DataGroupingType = keyof typeof dataGroupingTypes;

export interface CustomDataGroup {
  max: number;
  min: number;
}

export interface DataGroupingConfig {
  type: DataGroupingType;
  numberOfGroups?: number;
  customGroups: CustomDataGroup[];
}

export interface MapDataSetConfig {
  dataSet: DataSet;
  dataGrouping: DataGroupingConfig;
}

export interface MapConfig {
  dataSetConfigs: MapDataSetConfig[];
}

export interface ChartProps {
  data: TableDataResult[];
  meta: FullTableMeta;
  title?: string;
  titleType?: 'default' | 'alternative';
  alt: string;
  height: number;
  width?: number;
  axes: AxesConfiguration;
  legend?: LegendConfiguration;
  includeNonNumericData?: boolean;
  showDataLabels?: boolean;
  map?: MapConfig;
}

export interface StackedBarProps extends ChartProps {
  barThickness?: number;
  dataLabelPosition?: BarChartDataLabelPosition;
  stacked?: boolean;
}

export interface ChartCapabilities {
  canPositionLegendInline: boolean;
  canSize: boolean;
  canSort: boolean;
  hasGridLines: boolean;
  hasLegend: boolean;
  hasLegendPosition: boolean;
  hasLineStyle: boolean;
  hasReferenceLines: boolean;
  hasSymbols: boolean;
  requiresGeoJson: boolean;
  stackable: boolean;
}

export interface ChartDefinitionOptions {
  stacked?: boolean;
  height: number;
  width?: number;
  barThickness?: number;
  title?: string;
  titleType: 'default' | 'alternative';
  alt: string;
  includeNonNumericData?: boolean;
  showDataLabels?: boolean;
  dataLabelPosition?: BarChartDataLabelPosition | LineChartDataLabelPosition;
  // Map options
  boundaryLevel?: number;
  dataClassification?: DataGroupingType;
  dataGroups?: number;
}

export interface ChartDefinition {
  type: ChartType;
  name: string;
  capabilities: ChartCapabilities;
  options: {
    defaults?: Partial<ChartDefinitionOptions>;
  };
  legend: {
    defaults?: Partial<LegendConfiguration>;
  };
  axes: {
    [key in AxisType]?: ChartDefinitionAxis;
  };
}

export interface ChartDefinitionAxisCapabilities {
  canRotateLabel: boolean;
}

export interface ChartDefinitionAxis {
  axis?: Axis;
  id: string;
  title: string;
  type: AxisType;
  hide?: boolean;
  capabilities: ChartDefinitionAxisCapabilities;
  defaults?: NestedPartial<AxisConfiguration>;
  constants?: {
    groupBy?: AxisGroupBy;
  };
  referenceLineDefaults?: Partial<ReferenceLine>;
}

export const chartDefinitions: ChartDefinition[] = [
  lineChartBlockDefinition,
  verticalBarBlockDefinition,
  horizontalBarBlockDefinition,
  mapBlockDefinition,
  infographicBlockDefinition,
];
