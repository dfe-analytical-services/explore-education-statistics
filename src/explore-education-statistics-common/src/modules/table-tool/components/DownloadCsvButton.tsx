import ButtonText from '@common/components/ButtonText';
import {
  CategoryFilter,
  Filter,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import { FullTable } from '@common/modules/table-tool/types/fullTable';
import cartesian from '@common/utils/cartesian';
import React from 'react';
import { utils, writeFile } from 'xlsx';

export const getCsvData = (fullTable: FullTable): string[][] => {
  const { subjectMeta, results } = fullTable;
  const { indicators, filters, timePeriodRange, locations } = subjectMeta;

  const filterColumns = Object.values(filters).map(
    filterGroup => filterGroup.name,
  );

  const indicatorColumns = indicators.map(indicator => indicator.name);

  const columns = [
    'location',
    'location_code',
    'geographic_level',
    'time_period',
    ...filterColumns,
    ...indicatorColumns,
  ];

  const rows = cartesian<Filter>(
    locations,
    timePeriodRange,
    ...Object.values(filters).map(filterGroup => filterGroup.options),
  )
    .map(filterCombination => {
      const [location, timePeriod, ...filterOptions] = filterCombination as [
        LocationFilter,
        TimePeriodFilter,
        ...CategoryFilter[]
      ];

      const indicatorCells = indicators.map(indicator => {
        const matchingResult = results.find(result => {
          const { geographicLevel } = result;

          return (
            filterOptions.every(filter =>
              result.filters.includes(filter.value),
            ) &&
            result.location[geographicLevel] &&
            result.location[geographicLevel].code === location.value &&
            location.level === geographicLevel &&
            result.timePeriod === timePeriod.value
          );
        });

        if (!matchingResult) {
          return 'n/a';
        }

        return matchingResult.measures[indicator.value] ?? 'n/a';
      });

      if (indicatorCells.every(cell => cell === 'n/a')) {
        return [];
      }

      return [
        location.label,
        location.value,
        location.level,
        timePeriod.label.replace(/\//g, ''),
        ...filterOptions.map(column => column.label),
        ...indicatorCells,
      ];
    })
    .filter(row => row.length > 0);

  return [columns, ...rows];
};

export interface DownloadCsvButtonProps {
  fileName: string;
  fullTable: FullTable;
  onClick?: () => void;
}

const DownloadCsvButton = ({
  fileName,
  fullTable,
  onClick,
}: DownloadCsvButtonProps) => {
  return (
    <ButtonText
      onClick={() => {
        const workBook = utils.book_new();
        workBook.Sheets.Sheet1 = utils.aoa_to_sheet(getCsvData(fullTable));
        workBook.SheetNames[0] = 'Sheet1';

        writeFile(workBook, `${fileName}.csv`, {
          type: 'binary',
        });
        if (onClick) {
          onClick();
        }
      }}
    >
      Download the underlying data of this table (CSV)
    </ButtonText>
  );
};

export default DownloadCsvButton;
