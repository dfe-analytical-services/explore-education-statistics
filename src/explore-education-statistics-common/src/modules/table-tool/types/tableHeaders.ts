import { Filter } from '@common/modules/table-tool/types/filters';

export interface TableHeadersConfig {
  columns: Filter[];
  columnGroups: Filter[][];
  rows: Filter[];
  rowGroups: Filter[][];
}
