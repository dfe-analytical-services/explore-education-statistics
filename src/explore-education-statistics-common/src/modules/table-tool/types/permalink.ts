import { TableHeadersConfig } from '@common/modules/table-tool/types/tableHeaders';
import { TableDataQuery } from '@common/services/tableBuilderService';
import { FullTable } from '@common/modules/table-tool/types/fullTable';

interface PermalinkQuery extends TableDataQuery {
  configuration: {
    tableHeadersConfig: TableHeadersConfig;
  };
}

export interface Permalink {
  id: string;
  title: string;
  created: string;
  fullTable: FullTable;
  query: PermalinkQuery;
}
