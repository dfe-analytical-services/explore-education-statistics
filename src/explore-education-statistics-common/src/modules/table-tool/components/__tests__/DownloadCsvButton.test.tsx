import {
  CategoryFilter,
  Indicator,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import { FullTableMeta } from '@common/modules/table-tool/types/fullTable';
import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { WorkBook, writeFile } from 'xlsx';
import DownloadCsvButton, { getCsvData } from '../DownloadCsvButton';

jest.mock('xlsx', () => {
  const { utils } = jest.requireActual('xlsx');

  return {
    writeFile: jest.fn(),
    utils,
  };
});

describe('DownloadCsvButton', () => {
  const basicSubjectMeta: FullTableMeta = {
    geoJsonAvailable: false,
    publicationName: '',
    subjectName: '',
    footnotes: [],
    boundaryLevels: [],
    filters: {
      Characteristic: {
        name: 'characteristic',
        options: [
          new CategoryFilter({
            value: 'gender_female',
            label: 'Female',
            group: 'Gender',
            category: 'Characteristic',
          }),
        ],
      },
    },
    indicators: [
      new Indicator({
        label: 'Authorised absence rate',
        value: 'authAbsRate',
        unit: '%',
        name: 'sess_authorised_percent',
      }),
      new Indicator({
        label: 'Number of authorised absence sessions',
        value: 'authAbsSess',
        unit: '',
        name: 'sess_authorised',
      }),
    ],
    locations: [
      new LocationFilter({
        value: 'england',
        label: 'England',
        level: 'country',
      }),
    ],
    timePeriodRange: [
      new TimePeriodFilter({
        code: 'AY',
        year: 2015,
        label: '2015/16',
        order: 0,
      }),
    ],
  };

  test('calls `writeFile` when button is pressed', () => {
    const { getByText } = render(
      <DownloadCsvButton
        fileName="pupil-absence"
        fullTable={{
          subjectMeta: basicSubjectMeta,
          results: [],
        }}
      />,
    );

    fireEvent.click(
      getByText('Download the underlying data of this table (CSV)'),
    );

    const mockedWriteFile = writeFile as jest.Mock;

    expect(mockedWriteFile).toHaveBeenCalledTimes(1);

    const workbook = mockedWriteFile.mock.calls[0][0] as WorkBook;

    expect(workbook.Sheets.Sheet1.A1.v).toBe('location');
    expect(workbook.Sheets.Sheet1.B1.v).toBe('location_code');
    expect(workbook.Sheets.Sheet1.C1.v).toBe('geographic_level');
    expect(workbook.Sheets.Sheet1.D1.v).toBe('time_period');
    expect(workbook.Sheets.Sheet1.E1.v).toBe('characteristic');
    expect(workbook.Sheets.Sheet1.F1.v).toBe('sess_authorised_percent');
    expect(workbook.Sheets.Sheet1.G1.v).toBe('sess_authorised');

    expect(mockedWriteFile.mock.calls[0][1]).toBe('pupil-absence.csv');
  });

  describe('getCsvData', () => {
    test('contains full set of data', async () => {
      const data = getCsvData({
        subjectMeta: {
          ...basicSubjectMeta,
          filters: {
            ...basicSubjectMeta.filters,
            'School Type': {
              name: 'school_type',
              options: [
                new CategoryFilter({
                  value: 'school_primary',
                  label: 'State-funded primary',
                  category: 'School Type',
                }),
                new CategoryFilter({
                  value: 'school_secondary',
                  label: 'State-funded secondary',
                  category: 'School Type',
                }),
              ],
            },
          },
          locations: [
            ...basicSubjectMeta.locations,
            new LocationFilter({
              value: 'barnsley',
              label: 'Barnsley',
              level: 'localAuthority',
            }),
          ],
        },
        results: [
          {
            filters: ['gender_female', 'school_primary'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '111',
              authAbsSess: '222',
            },
          },
          {
            filters: ['gender_female', 'school_secondary'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '333',
              authAbsSess: '444',
            },
          },
          {
            filters: ['gender_female', 'school_primary'],
            timePeriod: '2015_AY',
            geographicLevel: 'localAuthority',
            location: {
              localAuthority: {
                code: 'barnsley',
                name: 'Barnsley',
              },
            },
            measures: {
              authAbsRate: '555',
              authAbsSess: '666',
            },
          },
          {
            filters: ['gender_female', 'school_secondary'],
            timePeriod: '2015_AY',
            geographicLevel: 'localAuthority',
            location: {
              localAuthority: {
                code: 'barnsley',
                name: 'Barnsley',
              },
            },
            measures: {
              authAbsRate: '777',
              authAbsSess: '888',
            },
          },
        ],
      });

      expect(data).toHaveLength(5);

      expect(data[0]).toHaveLength(8);

      expect(data[1]).toHaveLength(8);
      expect(data[1][6]).toBe('111');
      expect(data[1][7]).toBe('222');

      expect(data[2]).toHaveLength(8);
      expect(data[2][6]).toBe('333');
      expect(data[2][7]).toBe('444');

      expect(data[3]).toHaveLength(8);
      expect(data[3][6]).toBe('555');
      expect(data[3][7]).toBe('666');

      expect(data[4]).toHaveLength(8);
      expect(data[4][6]).toBe('777');
      expect(data[4][7]).toBe('888');

      expect(data).toMatchSnapshot();
    });

    test("contains n/a's if there are some matching results", () => {
      const data = getCsvData({
        subjectMeta: {
          ...basicSubjectMeta,
          filters: {
            ...basicSubjectMeta.filters,
            'School Type': {
              name: 'school_type',
              options: [
                new CategoryFilter({
                  value: 'school_primary',
                  label: 'State-funded primary',
                  category: 'School Type',
                }),
                new CategoryFilter({
                  value: 'school_secondary',
                  label: 'State-funded secondary',
                  category: 'School Type',
                }),
              ],
            },
          },
        },
        results: [
          {
            filters: ['gender_female', 'school_primary'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '111',
            },
          },
          {
            filters: ['gender_female', 'school_secondary'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsSess: '222',
            },
          },
        ],
      });

      expect(data).toHaveLength(3);

      expect(data[0]).toHaveLength(8);

      expect(data[1]).toHaveLength(8);
      expect(data[1][6]).toBe('111');
      expect(data[1][7]).toBe('n/a');

      expect(data[2]).toHaveLength(8);
      expect(data[2][6]).toBe('n/a');
      expect(data[2][7]).toBe('222');

      expect(data).toMatchSnapshot();
    });

    test("strips out rows with only n/a's", () => {
      const data = getCsvData({
        subjectMeta: {
          ...basicSubjectMeta,
          filters: {
            ...basicSubjectMeta.filters,
            'School Type': {
              name: 'school_type',
              options: [
                new CategoryFilter({
                  value: 'school_primary',
                  label: 'State-funded primary',
                  category: 'School Type',
                }),
                new CategoryFilter({
                  value: 'school_secondary',
                  label: 'State-funded secondary',
                  category: 'School Type',
                }),
              ],
            },
          },
        },
        results: [
          {
            filters: ['gender_female', 'school_primary'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '111',
              authAbsSess: '222',
            },
          },
        ],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(8);

      expect(data[1]).toHaveLength(8);
      expect(data[1][6]).toBe('111');
      expect(data[1][7]).toBe('222');

      expect(data).toMatchSnapshot();
    });

    test('returns only header if there are no results', () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [],
      });

      expect(data).toHaveLength(1);

      expect(data[0]).toHaveLength(7);

      expect(data).toMatchSnapshot();
    });

    test("contains n/a's if matching result but no matching indicator", () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [
          {
            filters: ['gender_female'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsSess: '111',
            },
          },
        ],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(7);

      expect(data[1]).toHaveLength(7);
      expect(data[1][5]).toBe('n/a');
      expect(data[1][6]).toBe('111');

      expect(data).toMatchSnapshot();
    });

    test('does not format values in any way', () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [
          {
            filters: ['gender_female'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '12300000',
              authAbsSess: '44255667.2356',
            },
          },
        ],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(7);

      expect(data[1]).toHaveLength(7);
      expect(data[1][5]).toBe('12300000');
      expect(data[1][6]).toBe('44255667.2356');

      expect(data).toMatchSnapshot();
    });

    test('can contain suppressed cells', () => {
      const data = getCsvData({
        subjectMeta: basicSubjectMeta,
        results: [
          {
            filters: ['gender_female'],
            timePeriod: '2015_AY',
            geographicLevel: 'country',
            location: {
              country: {
                code: 'england',
                name: 'England',
              },
            },
            measures: {
              authAbsRate: '13.4',
              authAbsSess: 'x',
            },
          },
        ],
      });

      expect(data).toHaveLength(2);

      expect(data[0]).toHaveLength(7);

      expect(data[1]).toHaveLength(7);
      expect(data[1][5]).toBe('13.4');
      expect(data[1][6]).toBe('x');

      expect(data).toMatchSnapshot();
    });
  });
});
