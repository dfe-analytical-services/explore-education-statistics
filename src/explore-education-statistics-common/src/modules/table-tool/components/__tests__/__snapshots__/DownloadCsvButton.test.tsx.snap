// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadCsvButton calls \`writeFile\` when button is pressed 1`] = `
Object {
  "SheetNames": Array [
    "Sheet1",
  ],
  "Sheets": Object {
    "Sheet1": Object {
      "!ref": "A1:H3",
      "A1": Object {
        "t": "s",
        "v": "location",
      },
      "A2": Object {
        "t": "s",
        "v": "England",
      },
      "A3": Object {
        "t": "s",
        "v": "England",
      },
      "B1": Object {
        "t": "s",
        "v": "location_code",
      },
      "B2": Object {
        "t": "s",
        "v": "england",
      },
      "B3": Object {
        "t": "s",
        "v": "england",
      },
      "C1": Object {
        "t": "s",
        "v": "geographic_level",
      },
      "C2": Object {
        "t": "s",
        "v": "country",
      },
      "C3": Object {
        "t": "s",
        "v": "country",
      },
      "D1": Object {
        "t": "s",
        "v": "time_period",
      },
      "D2": Object {
        "t": "s",
        "v": "201516",
      },
      "D3": Object {
        "t": "s",
        "v": "201516",
      },
      "E1": Object {
        "t": "s",
        "v": "characteristic",
      },
      "E2": Object {
        "t": "s",
        "v": "Female",
      },
      "E3": Object {
        "t": "s",
        "v": "Female",
      },
      "F1": Object {
        "t": "s",
        "v": "school_type",
      },
      "F2": Object {
        "t": "s",
        "v": "State-funded primary",
      },
      "F3": Object {
        "t": "s",
        "v": "State-funded secondary",
      },
      "G1": Object {
        "t": "s",
        "v": "sess_authorised_percent",
      },
      "G2": Object {
        "t": "s",
        "v": "123",
      },
      "G3": Object {
        "t": "s",
        "v": "234",
      },
      "H1": Object {
        "t": "s",
        "v": "sess_authorised",
      },
      "H2": Object {
        "t": "s",
        "v": "456",
      },
      "H3": Object {
        "t": "s",
        "v": "567",
      },
    },
  },
}
`;

exports[`DownloadCsvButton getCsvData can contain suppressed cells 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "13.4",
    "x",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains full set of data 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "school_type",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded primary",
    "123",
    "456",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded secondary",
    "234",
    "567",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains n/a's if matching result but no matching indicator 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "n/a",
    "123",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains n/a's if there are no results 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "n/a",
    "n/a",
  ],
]
`;

exports[`DownloadCsvButton getCsvData contains n/a's if there are some matching results 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "school_type",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded primary",
    "123",
    "456",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "State-funded secondary",
    "n/a",
    "n/a",
  ],
]
`;

exports[`DownloadCsvButton getCsvData does not format values in any way 1`] = `
Array [
  Array [
    "location",
    "location_code",
    "geographic_level",
    "time_period",
    "characteristic",
    "sess_authorised_percent",
    "sess_authorised",
  ],
  Array [
    "England",
    "england",
    "country",
    "201516",
    "Female",
    "12300000",
    "44255667.2356",
  ],
]
`;
